common/typecase.h

```diff
@@ -33,15 +33,26 @@ template <typename T> using get_signature = typename get_signature_impl<T>::type
 
 template <typename T> struct argtype_extractor : public argtype_extractor<decltype(&T::operator())> {};
 template <typename ClassType, typename ReturnType, typename ArgType>
-struct argtype_extractor<ReturnType (ClassType::*)(ArgType *) const>
-// we specialize for pointers to member function
-{
+struct argtype_extractor<ReturnType (ClassType::*)(ArgType) const> {
     using arg_type = ArgType;
 };
 
+namespace Sfinae {
+// Check if .tag() exists (from https://stackoverflow.com/a/9154394)
+// Indicates that type is a tagged pointer.
+template <class> struct sfinae_true : std::true_type {};
+
+template <class T> static auto test_has_tag(int) -> sfinae_true<decltype(std::declval<T>().tag())>;
+template <class> static auto test_has_tag(long) -> std::false_type;
+template <class T> struct has_tag : decltype(test_has_tag<T>(0)) {};
```

Comment:
I wish there was a less-indirect way to write this. I fussed around with an implementation with `std::conditional` for a bit, but this implementation makes sense to me.

Reply:
I just realized we have a `GENERATE_HAS_MEMBER` in `treemap.h` I might be able to move and repurpose if you don't mind a slightly dirtier diff.

Reply:
Turns out I can't use `GENERATE_HAS_MEMBER` because it creates a subclass of the passed in type (??) and I have classes marked final. Hmm, I wonder if I can update it a bit to use my version...

Reply:
I'm going to punt on this for now.